### ISSUE TITLE
documentation: be more verbose about the symbols which can be used

### ISSUE THREAD

**Issue by LinguList** [lingpy/lingpy#180]
@cysouw  mentioned that it would be nice to have some explicit list of symbols which lingpy accepts for computation. This can be done in a rather straightforward way, be it by dumping all of the benchmarks (which have symbols which are accepted) or by adding some more fine-grained explanation, along with a way to test what is tolerated and what is not.

Ultimately, it would be nice to have an online-app in which one pastes data, and which is then automatically processed for missing and existing symbols.

**Comment by cysouw** [comment]
@LinguList you make the list, I'll make the app :-)

**Comment by LinguList** [comment]
deal. I think this can even be automatized, so that the list is constantly being updated. @xrotwang was using this to automatically annotate to which degree a given dataset is linked to the concepticon...

**Comment by SimonGreenhill** [comment]
It would be useful to have a function to check a wordlist/data object for potential errors like this.

**Comment by LinguList** [comment]
Ideally, even in JS: you open the web-page, paste your data (or upload it from a drag-drop-button), and there you are: all types, tokens, and all errors are just listed, including frequencies and the like. It is even possible to do this from XLSX-spreadsheets. All it need is to find the time to program it. I estimate that it will take two hours for an initial application. But since JS is ugly to program, it is not the nicest time to spend, so if we had for a definite release of LingPy-2.4 around March/April next year, I guess it is realistic to have this listed as one of the milestones...

**Comment by cysouw** [comment]
I have implemented this functionality in R in the package qlcData <https://cran.r-project.org/web/packages/qlcData/index.html>. There is also an UNIX-executable as a front end for it here <https://github.com/cysouw/qlcPipe>. I have just started making a shiny-webinterface for this, so soon people can just go to a webpage and have their orthography checked <http://shiny.rstudio.com>. There is also a python implementation of the same algorithms, but you have to ask Steve Moran whether it works now. Steve and I are working on a complete write-up of the background of the problem of orthography checking - Just drop me a note if you want to read a readable (but not yet complete) draft.

**Comment by LinguList** [comment]
Nice! Reg. Stevens version: it is still included in LingPy proper, but does not quite work as nicely as the things you have now implemented (no modifications since middle of 2014). It was planned to spin it off from lingpy and have it run as a plugin, but I do not know about the progress there. At the moment, all orthography-parsing things are deprecated in LingPy and they will be completely taken out for v2.4, assuming that the functionalities can then be shifted to the pluging. Alternatively, with the R-version in the pipeline, people can also segmentize their data with the orthoprofiles and load them later into LingPy. This will make much more sense for heavily orthographic datasets, where it requires people anyway to sit and think of how they want to segmentize their data...

**Comment by xrotwang** [comment]
I think the most recent work on Steve's orthography tokenizer is this branch in Frank Nagel's fork: https://github.com/FrankNagel/orthotokenizer/tree/context_tokenizer

**Comment by xrotwang** [comment]
Maybe integrating all these efforts would be a nice track for a LingPy hackday.

**Comment by cysouw** [comment]
We have a complete rewrite of the work of Steve in Python, according to the new specifications. The one that is still in LingPy is complete outdated, and not powerful enough. So yes, please get that one out of LingPy, and use any new version as a plugin.

Still, what is really necessary in that model is that LingPy is very explicit about the character assumptions. Basically just a list of characters that is accepted should suffice, then we can all work around that spec. I would propose to be rather strict in that specification though: only allow things that are really used in LingPy. Any alternative representation should then be handled in the plugin, or outside of LingPy. So no U+222B "INTEGRAL" < ∫> instead of U+0283 "LATIN SMALL LETTER ESH" <ʃ> in the input :-).

You can of course include some workarounds inside LingPy, just as you have now, but that should be somehow 'private' for people that happen to know about such possibilities.

**Comment by LinguList** [comment]
Well, I don't think that we need to be strict here: it's just a matter of two options, one that is strict, and one that is not. As it also is at the moment: if you go for a very specific sound class model and define it yourself, you can make that one strict, but if you have loose models, they will accept and tolerate things accordingly. I don't see why we should switch to an only-strict standard, given that the majority of applications that have been done so far are less strict.

For the new version of the tokenizer, we will need some documentation as to how to use it as a plugin. This was also something that was missing in the old tokenizer. And I think that the web-based tool that @cysouw plans will be very important, since it is just so much more convenient to use interactive apps to check on things than to hack inside of text files.

**Comment by cysouw** [comment]
@LinguList: what I mean with "strict" is that I (for example) do not know which characters are really used inside a specific model. I sort-of guess that you have the following minimal (or 'strict' as I call it) specifications (just looking at converters inside LingPy):

SCA: A C B E D G I H K J M L O N P S R U T W Y 1 2 3 4 5 6 9 _ 
DOLGO: V K P H J M N S R T W + 1 _ 

But then, what does it mean that, for example, ASJP has it's own model? and likewise for all the other converters? The Burmish converter seems to use the SCA character-vocabulary. The Jäger converter seems to use the ASJP vocabulary.

ASJP: p b f v m w 8 t T d s z c n r l S Z C j H y k g x N q G h 7 L i e E I a u o + 1 2 3 4 5 6 _ 
Burmish: A C B E D G I H K J M L O N P S R U T W Y 1 2 3 4 5 6 9 _ 

Basically, if you decide that people can either use the SCA, DOLGO or ASJP representations internally, then that is a nice strict specification, and I (and others) can work with that by mapping any other strings to that specification.

**Comment by LinguList** [comment]
Ah, I see, then we agree. We currently have these three alphabets of DOLGO, ASJP, and SCA, and you could also determine your own mapping to any of them. We could add another one that is even broader, but it doesn't seem to make much sense, we also have CV now, which is only consonant vowel (and tone). 

And then, splitting the conversion from the internal alphabet seems to be a good idea, since it is more explicit: burmish would then use SCA, but have a specific converter from burmish to SCA. Yes, that sounds very useful to me.

**Comment by cysouw** [comment]
BTW: the ASJP vocabulary is not the one from the ASJP project. Find attached here two profiles for ASJP

- a 'draft' profile, which is all 48 Unicode Codepoint that occur in the current ASJP data
- a 'tailored' profile which lists all 793 [sic!] graphemes in the current ASJP data

[asjp_tailored_profile.txt](https://github.com/lingpy/lingpy/files/30671/asjp_tailored_profile.txt)
[asjp_draft_profile.txt](https://github.com/lingpy/lingpy/files/30672/asjp_draft_profile.txt)

**Comment by LinguList** [comment]
Yes, asjp is not identical with asjp-code. ASJP has a specific function in lingpy and some history: I use it specifically along with the weights of the Brown-et-al paper, wich are defined in the matrix and yield good results for dialectology. Since it's only internal conversion and the important point is the mapping to the sound-correspondence matrix, I changed some symbols which are needed for other things. Since asjp lacks tones, for example, I needed to add symbols for those, etc. 

Having this re-settled more transparently, both in LingPy and with additional tools, and in documentation, would be definitely useful.

BTW, in case you have not seen this, there is some documentation on the current idea behind the models and the like on the lingpy-doku [here](http://lingpy.org/reference/lingpy.data.html#lingpy.data.derive.compile_model).

**Comment by cysouw** [comment]
Maybe better not call it ASJP then... Maybe BSJP ;).

and: at some point somewhere add the following lists to the documentation, ideally with some explanation about the *meaning* of the symbols :-)

SCA: A C B E D G I H K J M L O N P S R U T W Y 1 2 3 4 5 6 9 _ 
DOLGO: V K P H J M N S R T W + 1 _
ASJP: p b f v m w 8 t T d s z c n r l S Z C j H y k g x N q G h 7 L i e E I a u o + 1 2 3 4 5 6 _

**Comment by LinguList** [comment]
well, for the meaning of dolgo and sca, they are given in my dissertation [at pages 123 and 124](http://dup.oa.hhu.de/244/1/DLS%20%28J.-M.%20List%29%2C%20Vol.%201_Open%20Access.pdf?page=145), where, by the way, you find other important aspects about sequence modeling in lingpy, which goes generally *beyond* simple strings, but uses multi-tiered sequence representations to handle aspects of context. In general, the closest description of the ideas behind most lingpy things can be found in the book. I did not have time to keep up with all aspects in the documentation, especially since in the end I was the only active user. But I'll definitely add that information in the update for v2.4.

**Comment by bambooforest** [comment]
Sorry, I'm a bit late to the party, but regarding messages regarding the
tokenizer way up there ^^

Yes, the version in Lingpy is severely outdated and it can be removed.

@LinguList decided here in #152 that the tokenizer should become a
stand-alone unit, which it long has, with a UNIX-executable, and as
@xrotwang points out (and he did the initial coding to make it CLI), the
most up-to-date version is at:

https://github.com/FrankNagel/orthotokenizer/tree/context_tokenizer

because it contains Frank's additional context code that @cysouw wanted
included.

Some work towards dropping the tokenizer from Lingpy was started in #158

**Comment by LinguList** [comment]
allright, so my question to @cysouw  and @bambooforest now: are these two versions of the orthoprifles, the r-version and the python-version identical in treatment and output? @cysouw's r-version has some new tweeks with intermediate classes, as I understand, is the same functionality given in the python-tokenizer?

**Comment by cysouw** [comment]
The plan is for the new R version and the new Python version to be identical. We are still testing whether this is the case.

However, basically they are parallel implementations of the same specification, so there might be detailed differences which we didn't realise up front. We are seeing this as a test whether the specifications are clear enough. When we find differences, we will have to become more precise in the specifications.

The specifications are part of the paper (book...) we are working on.

**Comment by bambooforest** [comment]
yes we are using the same format for the orthography profiles. although
both versions are under active development, so there might be differences
here and there still to be resolved as @cysouw continues to tweek, but the
idea is that they are essentially the same algorithms and processes in R
and Python (so people can choose their own poison)

**Comment by LinguList** [comment]
Well, in that case, you should consider adding a version in JavaScript. Why so? Because the Ortho-Profile stuff is not heavy in terms of computation, and one could immediately tweak things in a nice online interface. Trying to do these things in textfiles and running python scripts to check (or R-scripts) is something that not many people will think being worth the pain...

**Comment by xrotwang** [comment]
@LinguList debugging JS may also be considered not being worth the pain :) So I guess unless someone is scratching their own itch, creating a JS implementation may lead to a not-well-maintained, buggy piece of code.

**Comment by cysouw** [comment]
nope, it is quite a pain to get orthographies right, actually... We did two implementations, and I personally do not want to do a third implementation. The plan is the following:

- have implementations in R and Python for people working in those surroundings (R finished and available through the official R-channel, called CRAN, Python basically finished, needs final checking and bug-corrections)
- have a UNIX-executable interface to both of these versions for people using the command line (both finished, the R version is also working)
- have a web-interface for people not comfortable with programming (I am working on a "shiny" version of the R code, no plans for any Python or Javascript solution)

**Comment by LinguList** [comment]
I know it hurts to program JS, the web-based aspect and the possibility to work with things on the fly, is, however, something that will be important to first convince people that it's worth the pain of using it, and that it's easy to create the ortho-profiles for a new dataset. This was completely lacking in all orthocodes we had so far. But what you are working on at the moment sounds very promising, so looking forward to seeing the things developing further!
